File: src/features/test-card/test-card.service.ts
import { parseWorkItemId } from '../../shared/validation/work-item'
import type {
  DescriptionGenerator,
  GitContextReader,
  TestCardRepository,
  TestCardRepositoryFactory
} from '../../app/ports'
import type { ConfigService } from '../../infrastructure/config/config-service'
import type { CliOptions } from '../../app/cli.models'
import type { ProviderName } from '../../infrastructure/ai/ai.models'
import type { GitContext } from '../../infrastructure/git/git-context.models'
import type { AzurePullRequestChange, AzureWorkItem } from '../../azure'
import {
  buildTestCardPrompt,
  selectParentWorkItem,
  workItemText,
  TEST_CARD_SYSTEM_PROMPT,
  type TestCardContext,
  type TestCardPreparation
} from './test-card.models'
import { parseExamplesCount } from './test-card.validation'

export class TestCardService {
  constructor(
    private readonly config: ConfigService,
    private readonly git: GitContextReader,
    private readonly repositories: TestCardRepositoryFactory,
    private readonly generator: DescriptionGenerator
  ) {}

  async prepare(
    options: CliOptions,
    interactive: boolean,
    askForWorkItem: () => Promise<string | undefined>,
    report: (message: string) => void,
    onContext: (context: GitContext) => void
  ): Promise<TestCardPreparation> {
    const config = this.config.load(options)
    if (!config.azurePat.trim())
      throw new Error('O comando test requer AZURE_PAT ou AZURE_DEVOPS_PAT configurado.')
    if (options.create && !interactive)
      throw new Error('--create requer terminal interativo para confirmar o card.')

    const git = this.git.collect(options.source)
    onContext(git)
    if (!git.isAzureDevOps) throw new Error('O comando test requer um remote Git do Azure DevOps.')
    const repository = this.repositories.create(config, git)
    const pullRequestId = parseWorkItemId(options.pr, '--pr')
    const pullRequest = pullRequestId
      ? await repository.getPullRequest(git.azureProject, git.azureRepo, pullRequestId)
      : undefined

    let workItemId = parseWorkItemId(options.workItem, '--work-item')
    if (!workItemId && git.workItemId) workItemId = parseWorkItemId(git.workItemId, 'branch')
    if (!workItemId && pullRequest) {
      report('Resolvendo Work Item vinculado ao PR')
      const linkedIds = await repository.getPullRequestWorkItemIds(
        git.azureProject,
        git.azureRepo,
        pullRequest.pullRequestId
      )
      const linkedItems = await this.loadLinkedItems(repository, git.azureProject, linkedIds)
      workItemId = selectParentWorkItem(linkedItems)
    }
    if (!workItemId && interactive) {
      const value = await askForWorkItem()
      if (typeof value !== 'string') throw new Error('Operação cancelada.')
      workItemId = parseWorkItemId(value, 'Work Item')
    }
    if (!workItemId)
      throw new Error('Não foi possível resolver o Work Item pai; use --work-item explicitamente.')

    report(`Buscando Work Item #${workItemId}`)
    const workItem = await repository.getWorkItem(git.azureProject, workItemId)
    const changes = await this.loadChanges(repository, git, pullRequest)
    const examples = await this.loadExamples(repository, git.azureProject, options.examples)
    const context: TestCardContext = { git, workItem, pullRequest, changes, examples }
    return { config, context, repository, prompt: buildTestCardPrompt(context) }
  }

  generate(
    preparation: TestCardPreparation,
    report: (provider: ProviderName, model: string) => void
  ) {
    return this.generator.generate({
      config: preparation.config,
      system: TEST_CARD_SYSTEM_PROMPT,
      prompt: preparation.prompt,
      branch: `test-card/${preparation.context.workItem.id}`,
      report
    })
  }

  create(
    preparation: TestCardPreparation,
    input: Parameters<TestCardRepository['createTestCase']>[1]
  ) {
    return preparation.repository.createTestCase(preparation.context.git.azureProject, input)
  }

  updateParent(
    preparation: TestCardPreparation,
    effort?: number,
    realEffort?: number
  ): Promise<void> {
    return preparation.repository.updateWorkItemToTestQA(
      preparation.context.git.azureProject,
      preparation.context.workItem.id,
      effort,
      realEffort
    )
  }

  private async loadLinkedItems(
    repository: TestCardRepository,
    project: string,
    ids: number[]
  ) {
    const items: AzureWorkItem[] = []
    for (const id of ids) {
      try {
        items.push(await repository.getWorkItem(project, id))
      } catch {
        // Um link antigo ou sem permissão não bloqueia os demais.
      }
    }
    return items
  }

  private async loadChanges(
    repository: TestCardRepository,
    git: GitContext,
    pullRequest: TestCardContext['pullRequest']
  ): Promise<AzurePullRequestChange[]> {
    if (!pullRequest) return new Array<AzurePullRequestChange>()
    try {
      return await repository.getPullRequestChanges(git.azureProject, git.azureRepo, pullRequest)
    } catch {
      return new Array<AzurePullRequestChange>()
    }
  }

  private async loadExamples(
    repository: TestCardRepository,
    project: string,
    value: string | undefined
  ): Promise<string[]> {
    const count = parseExamplesCount(value)
    if (count === 0) return new Array<string>()
    try {
      const ids = await repository.queryTestCaseIds(project)
      const examples: string[] = []
      for (const id of ids.slice(0, count)) {
        try {
          const item = await repository.getWorkItem(project, id)
          const title = workItemText(item, 'System.Title')
          if (title) examples.push(`- #${id} ${title}`)
        } catch {
          // Um exemplo inacessível não impede a geração do card.
        }
      }
      return examples
    } catch {
      return new Array<string>()
    }
  }
}

File: lib/src/features/test_card/test_card_service.dart
import 'package:better_effect/better_effect.dart';

import '../../application/ai/description_generator.dart';
import '../../application/ai/description_models.dart';
import '../../app/app_effect.dart';
import '../../app/app_failure.dart';
import '../../app/cli_options.dart';
import '../../application/config/config_service.dart';
import '../../application/change_context/change_context_reader.dart';
import '../../application/terminal/terminal_ports.dart';
import 'test_card_models.dart';
import 'test_card_repository.dart';
import 'test_card_validation.dart';

abstract interface class TestCardService {
  AppEffect<TestCardPreparation> prepare(CliOptions options, bool interactive);

  AppEffect<GeneratedDescription> generate(
    TestCardPreparation preparation, {
    DescriptionReporter? report,
  });

  AppEffect<WorkItem> create(
    TestCardPreparation preparation,
    TestCardDraft input,
  );

  AppEffect<TestCardUpdate> updateParent(
    TestCardPreparation preparation, {
    num? effort,
    num? realEffort,
  });
}

final class TestCardFailure extends AppFailure {
  const TestCardFailure(String message) : super(message, 1);
}

final class TestCardServiceLive implements TestCardService {
  const TestCardServiceLive();

  @override
  AppEffect<TestCardPreparation> prepare(
    CliOptions options,
    bool interactive,
  ) => .result((use) async {
    final config = await use.unwrap(use<ConfigService>().load(options));
    if (config.azurePat.trim().isEmpty) {
      use.fail(
        const TestCardFailure(
          'O comando test requer AZURE_PAT ou AZURE_DEVOPS_PAT configurado.',
        ),
      );
    }
    if (options.create && !interactive) {
      use.fail(
        const TestCardFailure(
          '--create requer terminal interativo para confirmar o card.',
        ),
      );
    }

    final change = await use.unwrap(
      use<ChangeContextReader>().collect(options.source),
    );
    if (change.remote == null) {
      use.fail(
        const TestCardFailure(
          'O comando test requer um remote Git do Azure DevOps.',
        ),
      );
    }

    final repository = use<TestCardRepository>();
    final pullRequestId = options.pr == null
        ? null
        : int.parse(
            await use.result(parseWorkItemId(options.pr!.value, '--pr')),
          );
    final pullRequest = pullRequestId == null
        ? null
        : await use.unwrap(repository.getPullRequest(pullRequestId));

    String? workItemId = options.workItem?.value;
    if (workItemId != null) {
      workItemId = await use.result(parseWorkItemId(workItemId, '--work-item'));
    }
    if (workItemId == null && change.workItemId.isNotEmpty) {
      workItemId = await use.result(
        parseWorkItemId(change.workItemId, 'branch'),
      );
    }
    if (workItemId == null && pullRequest != null) {
      final linkedIds = await use.unwrap(
        repository.getPullRequestWorkItemIds(pullRequest.id),
      );
      final linkedItems = <WorkItem>[];
      for (final id in linkedIds) {
        final result = await use.unwrap(repository.getWorkItem(id).either());
        result.fold<void>(linkedItems.add, (_) {});
      }
      final parent = selectParentWorkItem(linkedItems);
      workItemId = parent?.toString();
    }
    if (workItemId == null && interactive) {
      final prompts = use<PromptPort>();
      final value = prompts.text(
        message: 'ID do Work Item pai',
        validate: validateWorkItemId,
      );
      if (value == null) use.fail(const TestCardFailure('Operação cancelada.'));
      workItemId = await use.result(parseWorkItemId(value, 'Work Item'));
    }
    if (workItemId == null) {
      use.fail(
        const TestCardFailure(
          'Não foi possível resolver o Work Item pai; use --work-item explicitamente.',
        ),
      );
    }

    final id = int.parse(workItemId);
    final workItem = await use.unwrap(repository.getWorkItem(id));
    final changes = <Change>[];
    if (pullRequest != null) {
      final result = await use.unwrap(
        repository.getPullRequestChanges(pullRequest).either(),
      );
      result.fold<void>(changes.addAll, (_) {});
    }

    final count = await use.result(parseExamplesCount(options.examples));
    final examples = <String>[];
    if (count > 0) {
      final idsResult = await use.unwrap(
        repository.queryTestCaseIds().either(),
      );
      List<int>? ids;
      idsResult.fold<void>((value) => ids = value, (_) {});
      for (final exampleId in (ids ?? const <int>[]).take(count)) {
        final itemResult = await use.unwrap(
          repository.getWorkItem(exampleId).either(),
        );
        itemResult.fold<void>((item) {
          final title = workItemText(item, 'System.Title');
          if (title.isNotEmpty) examples.add('- #$exampleId $title');
        }, (_) {});
      }
    }

    final context = TestCardContext(
      change: change,
      workItem: workItem,
      pullRequest: pullRequest,
      changes: List.unmodifiable(changes),
      examples: List.unmodifiable(examples),
    );
    return TestCardPreparation(
      config: config,
      context: context,
      prompt: buildTestCardPrompt(context),
    );
  });

  @override
  AppEffect<GeneratedDescription> generate(
    TestCardPreparation preparation, {
    DescriptionReporter? report,
  }) => .result((use) async {
    final generator = use<DescriptionGenerator>();
    return use.unwrap(
      generator.generate(
        config: preparation.config,
        system: testCardSystemPromptValue,
        prompt: preparation.prompt,
        branch: 'test-card/${preparation.context.workItem.id}',
        report: report,
      ),
    );
  });

  @override
  AppEffect<WorkItem> create(
    TestCardPreparation preparation,
    TestCardDraft input,
  ) => .result((use) async {
    final repository = use<TestCardRepository>();
    return use.unwrap(repository.createTestCase(input));
  });

  @override
  AppEffect<TestCardUpdate> updateParent(
    TestCardPreparation preparation, {
    num? effort,
    num? realEffort,
  }) => .result((use) async {
    final repository = use<TestCardRepository>();
    return use.unwrap(
      repository.updateWorkItemToTestQa(
        preparation.context.workItem.id,
        effort: effort,
        realEffort: realEffort,
      ),
    );
  });
}
